commit 67e5139b02518cbc83e5156e2f74a5d22f4d97a4
Author: Sage Weil <sweil@redhat.com>
Date:   Thu Oct 31 10:42:53 2019 -0500

    mgr/ActivePyModules: behave if a module queries a devid that does not exist
    
    The GIL reacquire was being handled in the lambda, but that was not
    getting called if the device didn't exist, leading to a crash.
    
    Add a trivial CLI test.
    
    Fixes: https://tracker.ceph.com/issues/42578
    Signed-off-by: Sage Weil <sage@redhat.com>

diff --git a/src/mgr/ActivePyModules.cc b/src/mgr/ActivePyModules.cc
index 27dbe175ee2..5a31ad0691d 100644
--- a/src/mgr/ActivePyModules.cc
+++ b/src/mgr/ActivePyModules.cc
@@ -341,10 +341,15 @@ PyObject *ActivePyModules::get_python(const std::string &what)
   } else if (what.size() > 7 &&
 	     what.substr(0, 7) == "device ") {
     string devid = what.substr(7);
-    daemon_state.with_device(devid, [&f, &tstate] (const DeviceState& dev) {
-        PyEval_RestoreThread(tstate);
-	f.dump_object("device", dev);
-      });
+    if (!daemon_state.with_device(
+	  devid,
+	  [&f, &tstate] (const DeviceState& dev) {
+	    PyEval_RestoreThread(tstate);
+	    f.dump_object("device", dev);
+	  })) {
+      // device not found
+      PyEval_RestoreThread(tstate);
+    }
     return f.get();
   } else if (what == "io_rate") {
     cluster_state.with_pgmap(
